ArcGISRuntimeSDKQt_SampleViewers/ArcGISRuntimeSDKQt_Samples/GAnalytics.cpp

```diff
@@ -0,0 +1,164 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#include "GAnalytics.h"
+
+#include <QDebug>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QNetworkAccessManager>
+#include <QNetworkReply>
+#include <QOperatingSystemVersion>
+#include <QQmlEngine>
+#include <QSettings>
+#include <QUuid>
+
+#if defined (GANALYTICS_API_KEY) && defined (GANALYTICS_STREAM_ID)
+GAnalytics::GAnalytics(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_apiSecret(QUOTE(GANALYTICS_API_KEY)),
+  m_measurementId(QUOTE(GANALYTICS_STREAM_ID))
+{
+}
+#endif // GANALYTICS_API_KEY && GANALYTICS_STREAM_ID
+
+GAnalytics::GAnalytics(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_apiSecret(""),
+  m_measurementId("")
+{
+}
+
+GAnalytics::~GAnalytics()
+{
+  endSession();
+}
+
+void GAnalytics::init()
+{
+  if (m_apiSecret == "" || m_measurementId == "")
+  {
+    m_telemetryEnabled = false;
+    setIsVisible(false);
+    return;
+  }
+
+  generateClientId();
+
+  if (!m_settings.contains("GAnalytics-telemetry-enabled"))
+  {
+    m_telemetryEnabled = true;
+    m_settings.setValue("GAnalytics-telemetry-enabled", m_telemetryEnabled);
+    setIsVisible(true);
+  }
+  else
+  {
+    m_telemetryEnabled = m_settings.value("GAnalytics-telemetry-enabled").toBool();
+  }
+
+  m_startTime = QDateTime::currentSecsSinceEpoch();
+
+  m_networkAccessManager = new QNetworkAccessManager();
+
+#ifdef CPP_VIEWER
+  m_defaultParameters.insert("sample_viewer_type", "Cpp");
+#else // QML_VIEWER
+  m_defaultParameters.insert("sample_viewer_type", "QML");
+#endif // CPP_VIEWER / QML_VIEWER
+
+  auto os = QOperatingSystemVersion::current();
+  m_defaultParameters.insert("operating_system", QString{"%1 %2.%3"}.arg(os.name(), os.majorVersion(), os.minorVersion()));
+}
+
+void GAnalytics::postEvent(const QString& eventName, QVariantMap parameters)
+{
+  if (!m_telemetryEnabled || m_apiSecret == "" || m_measurementId == "")
+    return;
+
+  int sessionTime = (QDateTime::currentSecsSinceEpoch() - m_startTime) * 1000;
+
+  // Google will not accept the POST request if the engagement time is 0 or undefined
+  if (sessionTime <= 0)
+    sessionTime = 1;
+
+  parameters.insert("engagement_time_msec", sessionTime);
+
+  QJsonObject event;
+  event.insert("name", eventName);
+  parameters.unite(m_defaultParameters);
+  event.insert("params", QJsonObject::fromVariantMap(parameters));
+
+  QJsonObject body;
+  body.insert("client_id", m_clientId);
+  body.insert("events", event);
+  const QUrl url(QString{"https://google-analytics.com/mp/collect?api_secret=%1&measurement_id=%2"}.arg(m_apiSecret, m_measurementId));
+
+  QNetworkRequest request(url);
+  request.setHeader(QNetworkRequest::ContentTypeHeader, "application, json");
+  QJsonDocument data(body);
+
+  m_networkAccessManager->post(request, data.toJson());
```

Comment:
Do you need to check for the response, or any errors? This seems suspicious to fire and forget here.

Disregard if you have alternate logic that doesn't need the QNetworkReply.

Reply:
No need to check for a response because we wouldn't do anything with it. This is intended to work silently in the background. Furthermore Google Analytics doesn't return anything from the request apart from Status 202.